### PR TITLE
repo: Use python toolchains

### LINKED ISSUE #20431: Newer release available `rules_python`: 0.8.0 (current: 0.7.0)
Package Name: rules_python@0.7.0
Current Version: 0.7.0@2022-03-11
Available Version: 0.8.0@2022-03-20
Upstream releases: https://github.com/bazelbuild/rules_python/releases

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message:

Also bumps `rules_python` -> 0.8.0 (Fix #20431 ) 

In order to get this to work and to accommodate the 3 stages of loading required (register toolchains, register pipsets, install pipsets) i have had to move things around a little and add a `python_dependencies.bzl` file

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
reWIPped this while i figure out how to resolve issues with python2 <> thrift libs

i have raised a ticket upstream here https://github.com/bazelbuild/rules_python/issues/668

i have tried just separating the thrift pipset and not specifying an interpreter - but that doesnt seem to work

im going to try forking `rules_python` and adding a py2 toolchain - not sure if that will work or how easy, but worth a try

another possible approach is to patch the thrift libs to work on python3 as @zuercher has attempted previously

**Comment by phlax** [comment]
arrgghh - checking the upstream python "standalone" repo - there is no py2 support there - so forking `rules_python` wont work without building a separate standalone py2

**Comment by phlax** [comment]
also fails for arm - altho a PR has just been raised to address this here https://github.com/bazelbuild/rules_python/pull/669

**Comment by mattklein123** [comment]
Regarding Thrift, it's a long running problem that that entire library set if unmaintained. I would really hate to hold up further progress given that. Should we just remove the tests that require PY2? cc @rgs1 @zuercher @envoyproxy/dependency-shepherds

**Comment by rgs1** [comment]
> Regarding Thrift, it's a long running problem that that entire library set if unmaintained. I would really hate to hold up further progress given that. Should we just remove the tests that require PY2? cc @rgs1 @zuercher @envoyproxy/dependency-shepherds

I thought we had a soft agreement that it was probably fine in some other issue. cc: @fishcakez

**Comment by mattklein123** [comment]
What was probably fine? Removing the tests that require PY2? If so, please do a PR. :)

**Comment by rgs1** [comment]
Yes, removing those tests :-)

I’ll clear it with Stephan and James and follow up with a PR.

**Comment by phlax** [comment]
this seems to be reliably failing due to disk space issues, which i think implies we are close to the space limit for this job

im going to add the disk cleanup for the x64 release job - if that works, ill raise a PR separately to add

cc @mattklein123

**Comment by phlax** [comment]
i think the only remaining blocker for this is arm support landing in `rules_python`

~relatedly, we probably want to update the DEPENDENCY_POLICY doc wrt to python deps (and perhaps with some info about dependency issues)

File: bazel/repositories_extra.bzl [hunk 1]
```diff
@@ -1,31 +1,17 @@
...
 
 # Envoy deps that rely on a first stage of dependency loading in envoy_dependencies().
-def envoy_dependencies_extra():
-    _python_deps()
+def envoy_dependencies_extra(python_version = PYTHON_VERSION):
     emsdk_deps()
     raze_fetch_remote_crates()
     wasmtime_fetch_remote_crates()
+    python_register_toolchains(
+        name = "python%s" % ("_".join(python_version.split(".")[:-1])),
```
Comment (phlax) [inline bazel/repositories_extra.bzl:15]: add a comment to show what this does

**Comment by phlax** [comment]
cc @wrowe i have managed to get this further by increasing the shards for the thrift test - it now seems to complete successfully

unfortunately its hitting another issue:

```console

   check.warn(importable)
  copying multidict\__init__.pyi -> build\lib.win-amd64-cpython-310\multidict
  copying multidict\_multidict.c -> build\lib.win-amd64-cpython-310\multidict
  copying multidict\py.typed -> build\lib.win-amd64-cpython-310\multidict
  creating build\lib.win-amd64-cpython-310\multidict\_multilib
  copying multidict\_multilib\defs.h -> build\lib.win-amd64-cpython-310\multidict\_multilib
  copying multidict\_multilib\dict.h -> build\lib.win-amd64-cpython-310\multidict\_multilib
  copying multidict\_multilib\istr.h -> build\lib.win-amd64-cpython-310\multidict\_multilib
  copying multidict\_multilib\iter.h -> build\lib.win-amd64-cpython-310\multidict\_multilib
  copying multidict\_multilib\pair_list.h -> build\lib.win-amd64-cpython-310\multidict\_multilib
  copying multidict\_multilib\views.h -> build\lib.win-amd64-cpython-310\multidict\_multilib
  running build_ext
  building 'multidict._multidict' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  [end of output]

```

**Comment by phlax** [comment]
by my reading of the code here https://github.com/envoyproxy/envoy-build-tools/blob/08c1f7edd91d5c891e8b94e87f0eafa6b4224fad/build_container/build_container_windows.ps1#L82-L98


```bash
# VS 2019 Build Tools
# Pinned to version 16.9.4 downloaded on 1/31/2022 via https://aka.ms/vs/16/release/vs_BuildTools.exe
DownloadAndCheck $env:TEMP\vs_BuildTools.exe `
                 https://download.visualstudio.microsoft.com/download/pr/9a26f37e-6001-429b-a5db-c5455b93953c/460d80ab276046de2455a4115cc4e2f1e6529c9e6cb99501844ecafd16c619c4/vs_BuildTools.exe `
                 460d80ab276046de2455a4115cc4e2f1e6529c9e6cb99501844ecafd16c619c4
# See: https://docs.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2019#c-build-tools
# Other dependent/required components including the "Microsoft.VisualStudio.Workload.MSBuildTools"
# are added by the installer, do not need to be listed below, and cannot be supressed.
echo @"
{
  "version": "1.0",
  "components": [
    "Microsoft.VisualStudio.Workload.VCTools",
    "Microsoft.VisualStudio.Component.Windows11SDK.22000",
    "Microsoft.VisualStudio.Component.VC.Tools.x86.x64",
  ]
}
```

we already have "Microsoft Visual C++ 14.0 or greater"

im guessing this is related to issues with the standalone windows python binary - ref:

- https://github.com/indygreg/python-build-standalone/issues/124
-  https://github.com/bazelbuild/rules_python/pull/694

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -721,10 +721,10 @@ REPOSITORY_LOCATIONS_SPEC = dict(
...
         project_name = "Python rules for Bazel",
         project_desc = "Bazel rules for the Python language",
-        project_url = "https://github.com/bazelbuild/rules_python",
-        version = "0.8.1",
-        sha256 = "cdf6b84084aad8f10bf20b46b77cb48d83c319ebe6458a18e9d2cebf57807cdd",
-        release_date = "2022-04-22",
+        project_url = "https://github.com/phlax/rules_python",
+        version = "7cd86e2dcc87dced66336fc684777df453391915",
+        sha256 = "fa6511b7ff9cc949a5bce5a4a1d628488151309f6e06b3997e5dfccba1ecdc79",
+        release_date = "2022-05-18",
```
Comment (wrowe) [inline bazel/repository_locations.bzl:727]: If we go with this workaround, please note which release of rules_python is expected to pick up the fix upstream, and leave the canonical origin url commented out

Comment (phlax) [reply bazel/repository_locations.bzl:727]: we dont need to - it was just me testing

it now works with the published version - there is an upstream PR to moved to the shared-msvc - and i think that also now works - i will test it before i deWIP this PR

needing to compile makes the issue kinda moot, at least for now tho, so i think we should be good to land with released version

**Comment by phlax** [comment]
@wrowe ive cleaned up the PR (apologies for squash/force push) and it should be ready for review (at last!)

**Comment by phlax** [comment]
cc @ggreenway it would be good to do at least a `./docs/build.sh` on mac with this PR to confirm nothing is broken there

**Comment by ggreenway** [comment]
> cc @ggreenway it would be good to do at least a `./docs/build.sh` on mac with this PR to confirm nothing is broken there

Build completed successfully on my mac. For reference, x86 running MacOS 12.4 (latest release).

**Comment by wrowe** [comment]
Looks like a fix format patch to apply and this should be gtg

**Review by wrowe — APPROVED** [review]
Much cleaner solution, TY!

**Pull request merged**